### PR TITLE
ReadBuffer: check for unread data on next()

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Try to check if there are situations in our code, where we implicitly discard data by calling next() on non-empty buffer.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@abyss7 We can do it intentionally to skip data and it's Ok.

**Comment by abyss7** [comment]
> @abyss7 We can do it intentionally to skip data and it's Ok.

Is there a point on trying to refactor this? It will potentially simplify the usage and debugging for multi-layer buffers' composition.

**Comment by alexey-milovidov** [comment]
> At least I'd like to add ignore(available()) where possible.

Yes, it will be much better.

File: src/IO/ReadBuffer.h [hunk 1]
```diff
@@ -197,7 +198,7 @@ class ReadBuffer : public BufferBase
       */
     virtual bool nextImpl() { return false; }
 
-    [[noreturn]] void throwReadAfterEOF()
+    [[noreturn]] static inline void throwReadAfterEOF()
```
Comment (alexey-milovidov) [inline src/IO/ReadBuffer.h:201]: I doubt about inline.

**Review by alexey-milovidov — APPROVED** [review]

File: src/IO/ReadBuffer.h (hunk 1 above)
Comment (alexey-milovidov) [reply src/IO/ReadBuffer.h]: Constructor of Exception can be heavy due to construction of std::string.
If this function is inlined to another small function, that small function may stop to be inlined by itself.

Note: it's just a wild guess.

**Pull request merged**